2011-08-15  Richard Sandiford  <richard.sandiford@linaro.org>

	gcc/
	* config/rs6000/rs6000.c (paired_expand_vector_init): Don't create
	CONST_VECTORs with symbolic elements.
	(rs6000_expand_vector_init): Likewise.

=== modified file 'gcc/config/rs6000/rs6000.c'
--- old/gcc/config/rs6000/rs6000.c	2011-07-27 18:17:15 +0000
+++ new/gcc/config/rs6000/rs6000.c	2011-08-16 08:59:36 +0000
@@ -5134,7 +5134,9 @@
   for (i = 0; i < n_elts; ++i)
     {
       x = XVECEXP (vals, 0, i);
-      if (!CONSTANT_P (x))
+      if (!(CONST_INT_P (x)
+	    || GET_CODE (x) == CONST_DOUBLE
+	    || GET_CODE (x) == CONST_FIXED))
 	++n_var;
     }
   if (n_var == 0)
@@ -5286,7 +5288,9 @@
   for (i = 0; i < n_elts; ++i)
     {
       x = XVECEXP (vals, 0, i);
-      if (!CONSTANT_P (x))
+      if (!(CONST_INT_P (x)
+	    || GET_CODE (x) == CONST_DOUBLE
+	    || GET_CODE (x) == CONST_FIXED))
 	++n_var, one_var = i;
       else if (x != CONST0_RTX (inner_mode))
 	all_const_zero = false;

